keps/sig-api-machinery/2309-idl/syntax.md

```diff
@@ -0,0 +1,635 @@
+   disallowed constructs (and many bad practices) in Kubernetes APIs
+   should not be possible to represent in the IDL.
+
+3. Kubernetes API recommendations should be the default behavior in the
+   language: when not explicitly overridden, users should get our
+   recommended behavior by default:
+
+   * For example, the recommendation is new list-maps have a merge key of
+     name.  When merge keys aren’t manually specified, this is the default
+     one.
+
+4. All information conveyed should be as structured as possible
+
+5. Naming in the IDL describes the serialized form.  Language bindings
+   (Go) are derived automatically from that.  Specific overrides for
+   legacy codebases occur via markers, since they are exceptions to the
+   rules.
+
+## Examples
+
+***Author's note***: the cannonical source for these examples is [in the
+repository][repo-example]; the example reproduced here is just to make it
+easier to comment.
+
+[repo-example]: https://github.com/DirectXMan12/idl/blob/main/kdlc/testdata/all-around.kdl
+
+```
+// double slashes add comments
+/* same with block comments */
+
+// Triple-slash comments denote documentation
+// If a line contains `# Words`, the following lines
+// will be placed in the corresponding OpenAPI documentation
+// field.
+
+/// core/v1 describes core k8s concepts
+group-version(group:"core", version:"v1") {
+    /// Pod is a group of related, colocated processes
+    ///
+    /// # Example
+    /// {"apiVersion": "v1", "kind": "Pod", "spec": {}}
+    kind Pod {
+        // types may be nested -- this is just a convenience for namespacing
+        // it does not imply privacy or anything
+
+        spec: Spec,
+        struct Spec {
+            // list-maps are ordered maps whose keys appear in their body.
+            // not specifying a key defaults to `key: name`.
+            // Go represents them as lists with specific merge keys, other
+            // languages may represent them other ways.
+            // They are automatically merged as maps in SMD.
+
+            // optional fields must list the keyword `optional` before the
+            // type.
+
+            /// list of volumes that can be mounted by containers belonging to
+            /// the pod
+            volumes: optional list-map(value: Volume),
+
+            // ...
+
+            dnsPolicy: optional(default: ClusterFirst) DNSPolicy,
+
+            // simple-maps are unordered maps from string-type values to
+            // primitive values
+            nodeSelector: optional simple-map(value: string),
+
+            // ...
+
+            @deprecated(msg: "use serviceAccountName instead")
+            serviceAccount: optional string,
+
+            // ...
+
+            readinessGates: optional list(value: ReadinessGate),
+            struct ReadinessGate { }
+
+            // create-only fields are immutable after creation (see immutability KEP)
+            containers: create-only list-map(value: Container),
```

Comment:
why `create-only` rather than `immutable` ?

Reply:
hmm... I think someone suggested it to me at some point.  `immutable` is fine too tbh -- I think they both communicate what's going on well :-)

Reply:
There is a distinction like
- can be created only on object creations
- can be added later, but never removed
- can be appended
- can be prepended
- can be reordered.